File: netwatch/netwatch.py
# ============================================================
#  What's Using My Internet? 🌐
#  Made by DoxNAF.online
#
#  This tool shows you exactly which apps on your computer
#  are connected to the internet right now!
#
#  HOW TO RUN:
#    1. Install Python from https://python.org  (it's free!)
#    2. Open your terminal / Command Prompt
#    3. Type:  pip install psutil
#    4. Type:  python netwatch.py
#    5. That's it! 🎉
# ============================================================

import time   # lets us wait between updates
import os     # lets us clear the screen
import sys    # lets us exit the program cleanly

# ── Step 1: Make sure psutil is installed ────────────────────
#
# psutil is a helper library that lets Python peek at what
# programs are running and what network connections they have.
#
# If it's not installed yet, we'll tell the user exactly how
# to get it — no confusing error messages!
#
try:
    import psutil
except ImportError:
    print()
    print("  ❌  Oops! A helper library called 'psutil' is missing.")
    print()
    print("  To install it, open your terminal and type:")
    print()
    print("      pip install psutil")
    print()
    print("  Then run this program again!")
    print()
    input("  Press Enter to close...")
    sys.exit(1)


# ── Helper: Clear the screen ──────────────────────────────────

def clear_screen():
    """
    Wipes the terminal so the display looks clean every update.
    Uses 'cls' on Windows and 'clear' on Mac/Linux.
    """
    os.system('cls' if os.name == 'nt' else 'clear')


# ── Helper: Turn a huge byte number into something readable ───

def bytes_to_readable(num_bytes):
    """
    Converts a big number of bytes into words a human understands.

    Examples:
      800          ->  "800 B"    (bytes — tiny!)
      8,192        ->  "8.0 KB"   (kilobytes — small file)
      8,388,608    ->  "8.0 MB"   (megabytes — like a song)
      8,589,934,592 -> "8.0 GB"  (gigabytes — like a movie)

    We use 1024 (not 1000) because that's how computers actually
    count memory and storage — 1 KB = 1024 bytes.
    """
    if num_bytes < 1_024:
        return f"{num_bytes} B"
    elif num_bytes < 1_024 ** 2:
        return f"{num_bytes / 1_024:.1f} KB"
    elif num_bytes < 1_024 ** 3:
        return f"{num_bytes / 1_024 ** 2:.1f} MB"
    else:
        return f"{num_bytes / 1_024 ** 3:.1f} GB"


def speed_to_readable(bytes_per_second):
    """
    Turns a speed (in bytes per second) into something like '2.5 MB/s'.
    Just calls bytes_to_readable and adds '/s' for 'per second'.
    """
    return bytes_to_readable(bytes_per_second) + "/s"


# ── Helper: Pick an emoji based on how many connections ───────

def pick_emoji(num_connections):
    """
    Returns a fun emoji that shows how busy an app is.

    More connections = more internet activity = more exciting emoji!
    """
    if num_connections >= 10:
        return "🔥"   # Super busy — probably downloading or streaming
    elif num_connections >= 5:
        return "📶"   # Pretty active
    elif num_connections >= 2:
        return "🌐"   # A couple of connections — normal browsing
    else:
        return "🔌"   # Just one quiet connection


# ── Main function: Keep watching the internet! ────────────────

def watch_internet():
    """
    This is the main loop. It checks what's using the internet
    every second and shows a live, refreshing display.

    Here's what it does each loop:
      1. Grab current network stats
      2. Compare to last time to get download/upload speeds
         (1 second apart = the difference IS bytes-per-second)
      3. Find every program with an active internet connection
      4. Print it all in a nice, easy-to-read layout
      5. Wait 1 second, then do it all again
    """

    # Take a starting snapshot of how many bytes have been
    # sent and received since the computer booted up.
    previous_stats = psutil.net_io_counters()

    # Wait 1 second so we have a "before" and "after" to compare.
    time.sleep(1)

    # ── The loop that runs forever (until Ctrl+C) ──
    # Each pass takes exactly 1 second (sleep at the bottom),
    # so  current - previous  gives a true bytes-per-second speed.
    while True:

        # Grab the latest network stats
        current_stats = psutil.net_io_counters()

        # Work out speed:
        #   current - previous = how many bytes moved in the last second
        download_speed = current_stats.bytes_recv - previous_stats.bytes_recv
        upload_speed   = current_stats.bytes_sent - previous_stats.bytes_sent

        # Wipe the screen so the output looks like it's updating in place
        clear_screen()

        # ── Print the title bar ──────────────────────────────
        print("=" * 62)
        print("  🌐  WHAT'S USING MY INTERNET?       DoxNAF.online")
        print("  Ctrl+C to quit at any time.")
        print("=" * 62)
        print()

        # ── Show live download / upload speeds ───────────────
        print(f"  ⬇️  Downloading right now:   {speed_to_readable(download_speed)}")
        print(f"  ⬆️  Uploading right now:     {speed_to_readable(upload_speed)}")
        print()

        # ── Find every app with an active internet connection ─
        print("-" * 62)
        print("  📋  Apps currently connected to the internet:")
        print("-" * 62)
        print()

        connected_apps = []

        # Go through every running program on the computer
        for process in psutil.process_iter(['pid', 'name']):
            try:
                # Ask psutil for this program's network connections
                connections = process.net_connections()

                # 'ESTABLISHED' means the connection is live and active
                # (other statuses like 'LISTEN' mean it's just waiting)
                active_connections = [
                    c for c in connections
                    if c.status == 'ESTABLISHED'
                ]

                # If the app has at least one active connection, add it
                if active_connections:
                    connected_apps.append({
                        'name':        process.info['name'] or '(unknown)',
                        'connections': len(active_connections),
                    })

            except (psutil.NoSuchProcess, psutil.AccessDenied, psutil.ZombieProcess):
                # These errors just mean we couldn't check a particular
                # process — that's normal, so we just skip it quietly.
                pass

        if not connected_apps:
            print("  😴  No apps appear to be connected right now.")
            print("      (You might be offline, or everything is idle.)")
        else:
            # Sort the list so the busiest apps show up at the top
            connected_apps.sort(key=lambda app: app['connections'], reverse=True)

            # Print each app — cap at 25 so it fits on one screen
            for app in connected_apps[:25]:
                name      = app['name']
                num_conn  = app['connections']
                emoji     = pick_emoji(num_conn)

                # Make the connection count sound natural
                if num_conn == 1:
                    conn_label = "1 connection"
                else:
                    conn_label = f"{num_conn} connections"

                # Pad the name so all the connection counts line up neatly
                print(f"  {emoji}  {name:<34}  {conn_label}")

            # If there were more than 25, let the user know
            if len(connected_apps) > 25:
                extra = len(connected_apps) - 25
                print(f"       ... and {extra} more")

        # ── Show total data used since computer start ─────────
        print()
        print("-" * 62)
        print()
        print(f"  📊  Total since your computer last started:")
        print(f"       Downloaded:  {bytes_to_readable(current_stats.bytes_recv)}")
        print(f"       Uploaded:    {bytes_to_readable(current_stats.bytes_sent)}")
        print()
        print("  🔄  Updating every second...   (Ctrl+C to quit)")
        print()

        # Save the current stats so we can compare again next loop
        previous_stats = current_stats

        # Wait 1 second before the next refresh.
        # Because we measure bytes over exactly 1 second, the difference
        # between current and previous stats IS the bytes-per-second speed.
        time.sleep(1)


# ── Entry point ───────────────────────────────────────────────
#
# This block only runs when you execute THIS file directly.
# (If another script imports this file, this block is skipped —
# that's what `if __name__ == "__main__"` does.)
#
if __name__ == "__main__":
    print()
    print("  🌐  Welcome to  WHAT'S USING MY INTERNET?")
    print()
    print("  This tool shows you which apps are using your internet")
    print("  connection right now, updated live every second.")
    print()
    print("  Starting up — hang on for one second...")
    print()

    try:
        watch_internet()
    except KeyboardInterrupt:
        # The user pressed Ctrl+C — exit gracefully with a nice message
        print()
        print()
        print("  👋  See ya! Thanks for using What's Using My Internet?")
        print("      DoxNAF.online")
        print()

File: netwatch/netwatch.c
/* ============================================================
 *  NetWatch — What's Using My Internet?  v3.0  (C edition)
 *  Made by DoxNAF.online
 *
 *  A fast, lightweight Windows console tool that shows
 *  exactly which apps are using the internet in real time.
 *
 *  Compile with MSVC:
 *    cl netwatch.c /link iphlpapi.lib ws2_32.lib
 *
 *  Compile with MinGW / GCC:
 *    gcc netwatch.c -o netwatch.exe -liphlpapi -lws2_32
 *
 *  Requires Windows Vista or later.
 * ============================================================ */

#define WIN32_LEAN_AND_MEAN
#define _CRT_SECURE_NO_WARNINGS

#include <windows.h>
#include <winsock2.h>
#include <ws2tcpip.h>
#include <iphlpapi.h>
#include <psapi.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#pragma comment(lib, "iphlpapi.lib")
#pragma comment(lib, "ws2_32.lib")
#pragma comment(lib, "psapi.lib")

/* ── Constants ─────────────────────────────────────────────── */

#define MAX_APPS  512
#define COL_WIDTH 38

/* ── Data types ─────────────────────────────────────────────── */

/* One entry in our app list */
typedef struct {
    char  name[MAX_PATH]; /* Process file name, e.g. "chrome.exe" */
    DWORD pid;            /* Windows Process ID                    */
    int   connections;    /* Number of ESTABLISHED TCP connections  */
} AppEntry;

/* Network byte counters (snapshot) */
typedef struct {
    ULONGLONG bytes_recv; /* Total bytes received across all NICs  */
    ULONGLONG bytes_sent; /* Total bytes sent across all NICs      */
} NetStats;

/* ── Helpers ────────────────────────────────────────────────── */

/*
 * bytes_to_readable — convert raw byte count into a human-readable
 * string like "4.2 MB" or "820 KB".
 */
static void bytes_to_readable(ULONGLONG bytes, char *buf, size_t buflen)
{
    if (bytes < 1024ULL)
        snprintf(buf, buflen, "%llu B", bytes);
    else if (bytes < 1024ULL * 1024)
        snprintf(buf, buflen, "%.1f KB", (double)bytes / 1024.0);
    else if (bytes < 1024ULL * 1024 * 1024)
        snprintf(buf, buflen, "%.1f MB", (double)bytes / (1024.0 * 1024.0));
    else
        snprintf(buf, buflen, "%.1f GB", (double)bytes / (1024.0 * 1024.0 * 1024.0));
}

/*
 * speed_to_readable — same as bytes_to_readable but appends "/s".
 */
static void speed_to_readable(ULONGLONG bps, char *buf, size_t buflen)
{
    char tmp[64];
    bytes_to_readable(bps, tmp, sizeof(tmp));
    snprintf(buf, buflen, "%s/s", tmp);
}

/*
 * pick_indicator — returns a short ASCII activity bar based on how
 * many connections the app has.
 */
static const char *pick_indicator(int conns)
{
    if (conns >= 10) return "[***]"; /* Very busy — streaming / downloading */
    if (conns >= 5)  return "[** ]"; /* Pretty active                        */
    if (conns >= 2)  return "[*  ]"; /* A couple of connections              */
    return "[.  ]";                  /* Single quiet connection               */
}

/*
 * get_process_name — fills `name` with the EXE filename (e.g.
 * "chrome.exe") for the given PID.  Returns 1 on success.
 */
static int get_process_name(DWORD pid, char *name, size_t namelen)
{
    HANDLE hProc;
    char   path[MAX_PATH] = {0};
    DWORD  size = MAX_PATH;
    char  *last_slash;

    hProc = OpenProcess(PROCESS_QUERY_LIMITED_INFORMATION, FALSE, pid);
    if (!hProc)
        return 0;

    if (!QueryFullProcessImageNameA(hProc, 0, path, &size)) {
        CloseHandle(hProc);
        return 0;
    }
    CloseHandle(hProc);

    /* Keep only the filename portion */
    last_slash = strrchr(path, '\\');
    strncpy(name, last_slash ? last_slash + 1 : path, namelen - 1);
    name[namelen - 1] = '\0';
    return 1;
}

/*
 * compare_apps — qsort comparator; sorts by connection count
 * descending so the busiest apps appear at the top.
 */
static int compare_apps(const void *a, const void *b)
{
    const AppEntry *x = (const AppEntry *)a;
    const AppEntry *y = (const AppEntry *)b;
    return y->connections - x->connections;
}

/*
 * clear_screen — erases the Windows console buffer without
 * spawning a separate "cls" process.
 */
static void clear_screen(void)
{
    HANDLE                     h = GetStdHandle(STD_OUTPUT_HANDLE);
    CONSOLE_SCREEN_BUFFER_INFO csbi;
    DWORD                      count;
    COORD                      origin = {0, 0};

    if (!GetConsoleScreenBufferInfo(h, &csbi))
        return;

    FillConsoleOutputCharacter(h, ' ',
        (DWORD)csbi.dwSize.X * csbi.dwSize.Y, origin, &count);
    FillConsoleOutputAttribute(h, csbi.wAttributes,
        (DWORD)csbi.dwSize.X * csbi.dwSize.Y, origin, &count);
    SetConsoleCursorPosition(h, origin);
}

/* ── Network statistics ─────────────────────────────────────── */

/*
 * get_net_stats — sums InOctets / OutOctets across all real
 * network adapters (skipping loopback and tunnel interfaces).
 * Returns 1 on success, 0 on failure.
 */
static int get_net_stats(NetStats *stats)
{
    MIB_IF_TABLE2 *table = NULL;
    ULONG i;

    if (GetIfTable2(&table) != NO_ERROR)
        return 0;

    stats->bytes_recv = 0;
    stats->bytes_sent = 0;

    for (i = 0; i < table->NumEntries; i++) {
        MIB_IF_ROW2 *row = &table->Table[i];

        /* Skip loopback and virtual tunnel adapters */
        if (row->Type == IF_TYPE_SOFTWARE_LOOPBACK) continue;
        if (row->Type == IF_TYPE_TUNNEL)            continue;

        stats->bytes_recv += row->InOctets;
        stats->bytes_sent += row->OutOctets;
    }

    FreeMibTable(table);
    return 1;
}

/* ── Connected-apps enumeration ─────────────────────────────── */

/*
 * add_or_inc_pid — looks up `pid` in the pid array; increments its
 * counter if found, otherwise appends a new entry.
 */
static void add_or_inc_pid(DWORD *pids, int *counts, int *n,
                            DWORD pid, int max)
{
    int j;
    for (j = 0; j < *n; j++) {
        if (pids[j] == pid) {
            counts[j]++;
            return;
        }
    }
    if (*n < max) {
        pids[*n]   = pid;
        counts[*n] = 1;
        (*n)++;
    }
}

/*
 * get_connected_apps — fills `apps` with every process that has at
 * least one ESTABLISHED TCP connection (IPv4 + IPv6).
 * Sets *app_count to the number of entries written.
 * Returns 1 on success, 0 on failure.
 */
static int get_connected_apps(AppEntry *apps, int *app_count)
{
    /* Temporary PID / count arrays */
    static DWORD pid_list[MAX_APPS];
    static int   conn_list[MAX_APPS];
    int          n = 0;
    int          i;
    DWORD        sz;

    *app_count = 0;

    /* ── IPv4 ─────────────────────────────────────────────── */
    sz = 0;
    GetExtendedTcpTable(NULL, &sz, FALSE, AF_INET,
                        TCP_TABLE_OWNER_PID_CONNECTIONS, 0);
    {
        MIB_TCPTABLE_OWNER_PID *t4 = (MIB_TCPTABLE_OWNER_PID *)malloc(sz);
        if (t4) {
            if (GetExtendedTcpTable(t4, &sz, FALSE, AF_INET,
                    TCP_TABLE_OWNER_PID_CONNECTIONS, 0) == NO_ERROR) {
                for (i = 0; i < (int)t4->dwNumEntries; i++) {
                    if (t4->table[i].dwState == MIB_TCP_STATE_ESTAB)
                        add_or_inc_pid(pid_list, conn_list, &n,
                                       t4->table[i].dwOwningPid, MAX_APPS);
                }
            }
            free(t4);
        }
    }

    /* ── IPv6 ─────────────────────────────────────────────── */
    sz = 0;
    GetExtendedTcpTable(NULL, &sz, FALSE, AF_INET6,
                        TCP_TABLE_OWNER_PID_CONNECTIONS, 0);
    {
        MIB_TCP6TABLE_OWNER_PID *t6 = (MIB_TCP6TABLE_OWNER_PID *)malloc(sz);
        if (t6) {
            if (GetExtendedTcpTable(t6, &sz, FALSE, AF_INET6,
                    TCP_TABLE_OWNER_PID_CONNECTIONS, 0) == NO_ERROR) {
                for (i = 0; i < (int)t6->dwNumEntries; i++) {
                    if (t6->table[i].dwState == MIB_TCP_STATE_ESTAB)
                        add_or_inc_pid(pid_list, conn_list, &n,
                                       t6->table[i].dwOwningPid, MAX_APPS);
                }
            }
            free(t6);
        }
    }

    /* ── Build AppEntry list ──────────────────────────────── */
    for (i = 0; i < n && *app_count < MAX_APPS; i++) {
        char name[MAX_PATH];
        AppEntry *e = &apps[*app_count];

        if (!get_process_name(pid_list[i], name, sizeof(name)))
            snprintf(name, sizeof(name), "(pid %lu)", (unsigned long)pid_list[i]);

        strncpy(e->name, name, sizeof(e->name) - 1);
        e->name[sizeof(e->name) - 1] = '\0';
        e->pid         = pid_list[i];
        e->connections = conn_list[i];
        (*app_count)++;
    }

    return 1;
}

/* ── Entry point ────────────────────────────────────────────── */

int main(void)
{
    NetStats  prev, curr;
    AppEntry  apps[MAX_APPS];
    int       app_count;
    int       limit, i;
    char      dl_buf[32], ul_buf[32];
    char      recv_buf[32], sent_buf[32];

    /* Use UTF-8 output on Windows 10+ */
    SetConsoleOutputCP(CP_UTF8);
    SetConsoleTitle("NetWatch -- What's Using My Internet?  |  DoxNAF.online");

    printf("\n");
    printf("  ==========================================================\n");
    printf("  WHAT'S USING MY INTERNET?   v3.0 (C edition)\n");
    printf("  Made by DoxNAF.online\n");
    printf("  ==========================================================\n\n");
    printf("  Starting up -- please wait one second...\n\n");

    if (!get_net_stats(&prev)) {
        fprintf(stderr, "  ERROR: Could not read network statistics.\n");
        printf("  Press Enter to exit...\n");
        getchar();
        return 1;
    }

    Sleep(1000); /* baseline measurement interval */

    while (1) {
        /* ── Take current snapshot ─────────────────────────── */
        if (!get_net_stats(&curr))
            break;

        /* speeds = bytes moved in the last ~1 second */
        ULONGLONG dl = (curr.bytes_recv > prev.bytes_recv)
                     ? curr.bytes_recv - prev.bytes_recv : 0;
        ULONGLONG ul = (curr.bytes_sent > prev.bytes_sent)
                     ? curr.bytes_sent - prev.bytes_sent : 0;

        /* ── Enumerate connected apps ──────────────────────── */
        get_connected_apps(apps, &app_count);
        if (app_count > 1)
            qsort(apps, app_count, sizeof(AppEntry), compare_apps);

        /* ── Render ────────────────────────────────────────── */
        clear_screen();

        printf("==============================================================\n");
        printf("  WHAT'S USING MY INTERNET?  (C edition)  |  DoxNAF.online\n");
        printf("  Press Ctrl+C to quit at any time.\n");
        printf("==============================================================\n\n");

        speed_to_readable(dl, dl_buf, sizeof(dl_buf));
        speed_to_readable(ul, ul_buf, sizeof(ul_buf));
        printf("  [v] Downloading:  %-16s\n", dl_buf);
        printf("  [^] Uploading:    %-16s\n\n", ul_buf);

        printf("--------------------------------------------------------------\n");
        printf("  Apps currently connected to the internet:\n");
        printf("--------------------------------------------------------------\n\n");

        if (app_count == 0) {
            printf("  (zzz)  No apps appear to be connected right now.\n");
            printf("         (You might be offline, or everything is idle.)\n");
        } else {
            limit = (app_count < 25) ? app_count : 25;
            for (i = 0; i < limit; i++) {
                const char *ind = pick_indicator(apps[i].connections);
                if (apps[i].connections == 1)
                    printf("  %s  %-*s  1 connection\n",
                           ind, COL_WIDTH, apps[i].name);
                else
                    printf("  %s  %-*s  %d connections\n",
                           ind, COL_WIDTH, apps[i].name, apps[i].connections);
            }
            if (app_count > 25)
                printf("       ... and %d more\n", app_count - 25);
        }

        bytes_to_readable(curr.bytes_recv, recv_buf, sizeof(recv_buf));
        bytes_to_readable(curr.bytes_sent, sent_buf, sizeof(sent_buf));

        printf("\n--------------------------------------------------------------\n\n");
        printf("  Total since your computer last started:\n");
        printf("       Downloaded:  %s\n", recv_buf);
        printf("       Uploaded:    %s\n\n", sent_buf);
        printf("  Updating every second...   (Ctrl+C to quit)\n\n");

        prev = curr;
        Sleep(1000);
    }

    return 0;
}
